README.md
XTurn - Xirsys TURN Server in Elixir
=====

This is an implementation of a TURN server in Elixir (based on the xstun server project).  It was originally written in Erlang and ported in 2014 when we migrated our other code.  It's never been in production and, indeed, needs more work for that.  However, it's a great little personal project and fun to work with.  It works nicely with WebRTC.

Supported Features
===

- TCP, UDP, TLS and DTLS supported
- Full TURN RFC5766 support (except rotating nonce)
- Full STUN RFC3489 support
- Simple user / pass storage with Web API interface
- Channel Binding / Data IS supported!
- WebRTC Data Channels ARE supported!

Setup
===
Open the `config.exs` file in `config`.  All options are there.

Logging
---
Logging sloooooows the server down.  For production quality (faster than Google's), drop the Logging level to `:error` or `:info`.  Keeping at `:debug` is fine for development, but will provide a degragation of service.

    config :logger,
      level: :debug,
      compile_time_purge_level: :debug

Ports
---
The listening ports should be set, next.  Standard ports are already set, but it can oftimes be beneficial to open on 80 and 443, too.  Make sure to specify `:secure` on known secure ports, which will enable SSL.

    config :xturn,
      authentication: %{required: true},
      permissions: %{required: false},
      realm: "xirsys.com",
      listen: [
                {:udp, '0.0.0.0', 3478},
                {:tcp, '0.0.0.0', 3478},
                {:udp, '0.0.0.0', 5349, :secure},
                {:tcp, '0.0.0.0', 5349, :secure}
              ],
      server_type: "turn",
      server_id: "turn.myserver.com",
      server_ip: {127, 0, 0, 1},
      server_local_ip: {0, 0, 0, 0},
      certs: [
               {:certfile, "certs/server.crt"},
               {:keyfile, "certs/server.key"}
             ]

*authentication*: specifying required as `true` will prevent connections without a valid user and password in the user store

*permissions*: TURN usually requires a `create permissions` call.  Setting requireed to false will allow connections without permissions being set.

*server_ip*: this is the public IP of your server.  Not all server setups make this aware to the app, so it's necessary to set this manually (for now).

*server_local_ip*: this is the internal IP to bind sockets to.  Again, this may be temporary.  You still need to set the IP in the individual socket listeners, too.

Note that `server_type` is a Xirsys thing and can be ignored.

Future Plans
===

- Create a rotating nonce
- Get a decent user credential store working with decent timeout capability (it's a little limited at the moment).
- Get RTP and RTCP working with a new MCU or SFU functionality
- Implement stream recording to file
- Implement third party streaming server connectivity
- Full support for IPv6
- TCP Allocations (connect command)

Donations
===
Developing XTurn takes time (mostly personal), so donations are welcome. Donations can be sent via PayPal:

![Donation QR Code](./docs/assets/images/qrcode.png)

Alternatively, if you would like to contribute, please feel free to let us know.

Changelog
===
Moved to separate document

Contact
===
For questions or suggestions, please email experts@xirsys.com

Copyright
===

Copyright (c) 2013 - 2020 Xirsys LLC

All rights reserved.

XTurn is licensed by Xirsys under the Apache License, Version 2.0. See LICENSE for the full license text.

CHANGELOG.md
# Changelog

07-01-2021 - Added Docker and Kubernetes files with support in the application for auto-scaling.

24-03-2020 - Updated to use the latest version of xturn-sockets, thereby simplifying package receipt and dispatch.

22-07-2019 - Hash key before sending to STUN.  Update dependency versions.  Fixed hash key correctly going to integrity binding.

19-07-2019 - Package for deploy to hex.pm

12-07-2019 - Fixed issue with integrity binding failing due to non-hashing of keys.

05-07-2019 - Simplified repo by moving common functionality to external libraries. XTurn repo now focuses on TURN specific functionality. Also added data receipt hooks for plugin capabilities.

22-01-2019 - Updated to work with latest XMediaLib

19-11-2018 - Simplify Pipelines module, moving pipeline descriptors to the config file. Also added [Getting Started](getting-started.md) page, describing testing the server.

30-10-2018 - Extract actions to module based pipeline

12-07-2018 - Externalised XMediaLib as a separate library

04-07-2018 - Aside from some cleanup, client calls were short-cicuited through direct passing of the client socket ref, rather than the GenServer pid

02-07-2018 - Get working with test.webrtc.org

26-06-2018 - Add DTLS support

21-09-2014 - Convert to Elixir

14-12-2013 - Initial working implementation in Erlang


config/config.exs
use Mix.Config

config :logger,
  level: :debug

config :xturn,
  authentication: %{required: false, username: "guest", credential: "guest"},
  # for some reason, turnutils_uclient doesn't set permissions for DTLS
  permissions: %{required: false},
  realm: "xturn.me",
  listen: [
    {:udp, '0.0.0.0', 80},
    {:tcp, '0.0.0.0', 80},
    {:udp, '0.0.0.0', 3478},
    {:tcp, '0.0.0.0', 3478},
    {:udp, '0.0.0.0', 443, :secure},
    {:tcp, '0.0.0.0', 443, :secure},
    {:udp, '0.0.0.0', 5349, :secure},
    {:tcp, '0.0.0.0', 5349, :secure}
  ],
  server_type: "turn",
  server_id: "xturn.me",
  server_ip: {0, 0, 0, 0},
  server_local_ip: {127, 0, 0, 1},
  certs: [
    cacertfile: 'certs/xturn.ca.pem',
    certfile: 'certs/xturn.cert.pem',
    keyfile: 'certs/xturn.key.pem'
  ],
  client_hooks: [],
  peer_hooks: [],
  pipes: %{
    allocate: [
      Xirsys.XTurn.Actions.HasRequestedTransport,
      Xirsys.XTurn.Actions.NotAllocationExists,
      Xirsys.XTurn.Actions.Authenticates,
      Xirsys.XTurn.Actions.Allocate
    ],
    refresh: [
      Xirsys.XTurn.Actions.Authenticates,
      Xirsys.XTurn.Actions.Refresh
    ],
    channelbind: [
      Xirsys.XTurn.Actions.Authenticates,
      Xirsys.XTurn.Actions.ChannelBind
    ],
    createperm: [
      Xirsys.XTurn.Actions.Authenticates,
      Xirsys.XTurn.Actions.CreatePerm
    ],
    send: [
      Xirsys.XTurn.Actions.SendIndication
    ],
    channeldata: [
      Xirsys.XTurn.Actions.ChannelData
    ]
  }

if Mix.env() == "dev" do
  config :peerage,
    via: Peerage.Via.List,
    node_list: [:"xturn@127.0.0.1"],
    log_results: false
end